src/crimson/common/operation.h

```diff
@@ -523,26 +537,32 @@ class PipelineHandle {
   template <typename OpT, typename T>
   seastar::future<>
   enter(T &stage, typename T::BlockingEvent::template Trigger<OpT>&& t) {
+    auto ret = enter_maybe_sync<OpT, T>(stage, std::move(t));
+    if (ret.has_value()) {
+      return std::move(ret.value());
+    } else {
+      return seastar::now();
```

Comment:
Why not just return a ready future from enter_maybe_sync?

Reply:
Upon leaving a concurrent phase and entering an exclusive phase, `enter_maybe_sync()` can return a non-ready future.

So its "maybe" sync.

Reply:
I mean, if ret were a ```seastar::future<>``` rather than ```std::optional<seastar::future<>>``` and you unconditionally returned it, it seems like the behavior would be identical?

Reply:
I suppose the recover_missing_snaps portion is the important one here.

Reply:
I'm not sure if I get you question correctly.

For example, if ret were a `seastar::future<>` in the above example (leaving a concurrent phase and entering an exclusive phase), returning it to the caller should be correct to preserve ordering.

Wating for that future to be resolved means that the request has left the concurrent phase in order, and has already enterred the follow-up exclusive phase in the correct order.

Reply:
> Why not just return a ready future from enter_maybe_sync?

Hm, probably I'm answering in the wrong direction.

`enter_maybe_sync()` is supposed to return non-future `nullopt` to support `lock_obc()` implementations.

There are many other existing users of `enter()` that don't expect `std::option<future>`, and I think it is unnecessary to adjust them, so `enter()` wraps `enter_maybe_sync()` to always return a future like before.